File: software/zyarm_sdk/python/src/zyarm_sdk/retarget.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Sequence, Tuple

from .safety import validate_apply_mask, validate_positions
from .types import JOINT_COUNT, TeleopAction


@dataclass(frozen=True)
class RetargetConfig:
    signs: Tuple[float, ...] = (1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0)
    offsets: Tuple[float, ...] = (0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    deadband: Tuple[float, ...] = (0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    min_positions: Optional[Tuple[float, ...]] = None
    max_positions: Optional[Tuple[float, ...]] = None
    max_delta: Optional[Tuple[float, ...]] = None
    apply_mask: Tuple[bool, ...] = (True, True, True, True, True, True, True)


class Retargeter:
    def __init__(self, config: Optional[RetargetConfig] = None) -> None:
        self.config = config or RetargetConfig()
        validate_positions(self.config.signs)
        validate_positions(self.config.offsets)
        validate_positions(self.config.deadband)
        validate_apply_mask(self.config.apply_mask)
        self._last: Optional[list[float]] = None

    @property
    def apply_mask(self) -> Tuple[bool, ...]:
        return tuple(bool(value) for value in self.config.apply_mask)

    def apply(self, action: TeleopAction) -> list[float]:
        source = validate_positions(action.positions)
        output: list[float] = []
        for index, value in enumerate(source):
            mapped = value * self.config.signs[index] + self.config.offsets[index]
            if self._last is not None and abs(mapped - self._last[index]) < self.config.deadband[index]:
                mapped = self._last[index]
            if self.config.max_delta is not None and self._last is not None:
                delta = mapped - self._last[index]
                step = self.config.max_delta[index]
                if delta > step:
                    mapped = self._last[index] + step
                elif delta < -step:
                    mapped = self._last[index] - step
            if self.config.min_positions is not None:
                mapped = max(mapped, self.config.min_positions[index])
            if self.config.max_positions is not None:
                mapped = min(mapped, self.config.max_positions[index])
            output.append(mapped)
        self._last = list(output)
        return output

File: software/zyarm_sdk/cpp/src/retarget.cpp
#include "zyarm_sdk/retarget.hpp"

#include <algorithm>
#include <cmath>

namespace zyarm_sdk
{

Retargeter::Retargeter() = default;

Retargeter::Retargeter(RetargetConfig config) : config_(config) {}

JointArray Retargeter::apply(const TeleopAction & action)
{
  JointArray out{};
  for (std::size_t index = 0; index < kJointCount; ++index) {
    double value = action.positions[index] * config_.signs[index] + config_.offsets[index];
    if (last_.has_value() && std::abs(value - (*last_)[index]) < config_.deadband[index]) {
      value = (*last_)[index];
    }
    if (last_.has_value() && config_.max_delta.has_value()) {
      const double delta = value - (*last_)[index];
      const double step = (*config_.max_delta)[index];
      if (delta > step) {
        value = (*last_)[index] + step;
      } else if (delta < -step) {
        value = (*last_)[index] - step;
      }
    }
    if (config_.min_positions.has_value()) {
      value = std::max(value, (*config_.min_positions)[index]);
    }
    if (config_.max_positions.has_value()) {
      value = std::min(value, (*config_.max_positions)[index]);
    }
    out[index] = value;
  }
  last_ = out;
  return out;
}

std::array<bool, kJointCount> Retargeter::apply_mask() const
{
  return config_.apply_mask;
}

}  // namespace zyarm_sdk
